data/abi/abi_encode_test.go

```diff
@@ -0,0 +1,1016 @@
+			}
+			// (2^[size] - 1) / (10^[precision]) test
+			ufixedLargestValue, err := MakeUfixed(largest, uint16(size), uint16(precision))
+			require.NoError(t, err, "make largest ufixed fail")
+			ufixedLargestEncode, err := ufixedLargestValue.Encode()
+			require.NoError(t, err, "largest ufixed encode error")
+			require.Equal(t, largest.Bytes(), ufixedLargestEncode,
+				"encode ufixed largest do not match with expected")
+		}
+	}
+
+	// encoding test for address, since address is 32 byte, it can be considered as 256 bit uint
+	// randomly generate 1000 uint256 and make address values, check if encoded value match with expected
+	upperLimit := big.NewInt(0).Lsh(big.NewInt(1), 256)
+	for i := 0; i < 1000; i++ {
+		randomAddrInt, err := rand.Int(rand.Reader, upperLimit)
+		require.NoError(t, err, "cryptographic random int init fail")
+
+		rand256Bytes := randomAddrInt.Bytes()
+		addrBytesExpected := make([]byte, 32-len(rand256Bytes))
+		addrBytesExpected = append(addrBytesExpected, rand256Bytes...)
+
+		var addrBytes [32]byte
+		copy(addrBytes[:], addrBytesExpected[:32])
+
+		addressValue := MakeAddress(addrBytes)
+		addrBytesActual, err := addressValue.Encode()
+		require.NoError(t, err, "address encode fail")
+		require.Equal(t, addrBytesExpected, addrBytesActual, "encode addr not match with expected")
+	}
+
+	// encoding test for bool values
+	for i := 0; i < 2; i++ {
+		boolValue := MakeBool(i == 1)
+		boolEncode, err := boolValue.Encode()
+		require.NoError(t, err, "bool encode fail")
+		expected := []byte{0x00}
+		if i == 1 {
+			expected = []byte{0x80}
+		}
+		require.Equal(t, expected, boolEncode, "encode bool not match with expected")
+	}
+
+	// encoding test for byte values
+	for i := 0; i < (1 << 8); i++ {
+		byteValue := MakeByte(byte(i))
+		byteEncode, err := byteValue.Encode()
+		require.NoError(t, err, "byte encode fail")
+		expected := []byte{byte(i)}
+		require.Equal(t, expected, byteEncode, "encode byte not match with expected")
+	}
+
+	// encoding test for string values, since strings in ABI contain utf-8 symbols
+	// we use `gobberish` to generate random utf-8 symbols
+	// randomly generate utf-8 str from length 1 to 100, each length draw 10 random strs
+	// check if encoded ABI str match with expected value
+	for length := 1; length <= 100; length++ {
+		for i := 0; i < 10; i++ {
+			// generate utf8 strings from `gobberish` at some length
+			utf8Str := gobberish.GenerateString(length)
+			strValue := MakeString(utf8Str)
+			// since string is just type alias of `byte[]`, we need to store number of bytes in encoding
+			utf8ByteLen := len([]byte(utf8Str))
+			lengthBytes := make([]byte, 2)
+			binary.BigEndian.PutUint16(lengthBytes, uint16(utf8ByteLen))
+			expected := append(lengthBytes, []byte(utf8Str)...)
+
+			strEncode, err := strValue.Encode()
+			require.NoError(t, err, "string encode fail")
+			require.Equal(t, expected, strEncode, "encode string not match with expected")
+		}
+	}
```

Comment:
The reason I used `gobberish` library is that, in ABI spec, strings are following utf-8 encoding, which means a string element may be more than 1 byte. To test this, `gobberish` provides random utf-8 strings to check if encoding/decoding of ABI strings provide a round-trip.